test/celeritas/track/TrackSort.test.cc

```diff
@@ -0,0 +1,179 @@
+#include "../global/StepperTestBase.hh"
+#include "celeritas_test.hh"
+
+namespace celeritas
+{
+namespace test
+{
+
+using celeritas::units::MevEnergy;
+
+#define TestTrackPartitionEm3Stepper \
+    TEST_IF_CELERITAS_GEANT(TestTrackPartitionEm3Stepper)
+class TestTrackPartitionEm3Stepper : public TestEm3Base, public StepperTestBase
+{
+  protected:
+    auto build_init() -> SPConstTrackInit override
+    {
+        TrackInitParams::Input input;
+        input.capacity = 4096;
+        input.max_events = 4096;
+        input.track_order = TrackOrder::partition_status;
+        return std::make_shared<TrackInitParams>(input);
+    }
+
+  public:
+    std::vector<Primary>
+    make_primaries_with_energy(size_type count, MevEnergy energy) const
+    {
+        Primary p;
+        p.particle_id = this->particle()->find(pdg::electron());
+        CELER_ASSERT(p.particle_id);
+        p.energy = energy;
+        p.track_id = TrackId{0};
+        p.position = {-22, 0, 0};
+        p.direction = {1, 0, 0};
+        p.time = 0;
+
+        std::vector<Primary> result(count, p);
+        for (auto i : range(count))
+        {
+            result[i].event_id = EventId{i};
+        }
+        return result;
+    }
+
+    //! Make 10GeV electrons along +x
+    std::vector<Primary> make_primaries(size_type count) const override
+    {
+        return this->make_primaries_with_energy(count, MevEnergy{10000});
+    }
+
+    size_type max_average_steps() const override
+    {
+        return 100000;  // 8 primaries -> ~500k steps, be conservative
+    }
+
+    GeantPhysicsOptions build_geant_options() const override
+    {
+        auto opts = TestEm3Base::build_geant_options();
+        opts.msc = MscModelSelection::none;
+        return opts;
+    }
+};
+
+TEST_F(TestTrackPartitionEm3Stepper, host_is_partitioned)
+{
+    size_type num_primaries = 8;
+    size_type num_tracks = 128;
+
+    Stepper<MemSpace::host> step(this->make_stepper_input(num_tracks));
+
+    // Initialize some primaries and take a step
+    auto primaries = this->make_primaries(num_primaries);
+    step(make_span(primaries));
+    auto check_is_partitioned = [&step]() {
+        auto span = step.core_data()
+                        .states.track_slots[AllItems<TrackSlotId::size_type>{}];
+        return std::is_partitioned(
+            span.begin(),
+            span.end(),
+            [&status = step.core_data().states.sim.status](
+                auto const track_slot) {
+                return status[TrackSlotId{track_slot}] == TrackStatus::alive;
+            });
+    };
+    // we partition at the start of the step so we need to explictly partition
+    // again after a step before checking
+    for (auto i = 0; i < 10; ++i)
+    {
+        detail::partition_tracks_by_status(step.core_data().states);
+        EXPECT_TRUE(check_is_partitioned()) << "Track slots are not "
+                                               "partitioned by status";
+        step();
+    }
```

Comment:
We do the partitioning at the beginning of the step; I suppose that some kernels executed later in the step can update the status of a track, so the assertion will fail if I don't explicitly call the partitioning of tracks again.

Reply:
That's to be expected I think; since right now all our kernels still have the `if (status != alive) return;` filter.